### ISSUE TITLE
Maximum call stack size exceeded

### ISSUE THREAD

**Issue by dlebedynskyi** [avajs/eslint-plugin-ava#179]
After updating to latest eslint > v4.7.0, we start to get following errors 
```
Maximum call stack size exceeded
RangeError: Maximum call stack size exceeded
at cloneOf (<Project>/node_modules/espurify/lib/clone-ast.js:54:22)
    at cloneObject (<Project>/node_modules/espurify/lib/clone-ast.js:49:26)
```

it seem to track down to eslint-plugin-ava. After removing plugin "ava" and "plugin:ava/recommended" - all works.

**Comment by twada** [comment]
`espurify`'s author here. Seems that it's a problem of [espurify](https://github.com/estools/espurify). Thank you for reporting.

Can I see your test code?
Or would you create a small reproduction case and file a issue to [espurify](https://github.com/estools/espurify) project?

**Comment by not-an-aardvark** [comment]
Hello :wave:

This was caused by https://github.com/eslint/eslint/issues/9122, which was reverted in `eslint@4.7.2`, so the problem with `eslint-plugin-ava` should now be fixed. (This issue was also reported in https://github.com/eslint/eslint/issues/9331.)

To summarize https://github.com/eslint/eslint/issues/9122: ESLint adds a `parent` property to each AST node as it traverses the AST, pointing to the parent of the given node. Previously, this happened at the same time as ESLint rule listeners were called, so at the start of the traversal, there would be no `parent` properties. In ESLint v4.7.0, this was changed so that the `parent` property would be added before any rules were run.

This is generally an improvement for ESLint rule authors, because they don't have to worry about whether a `parent` property on a given node has been set at any given point. However a side-effect is that the AST object contains circular references when rules initially receive it, which seems to cause a stack overflow in `espurify`.

We've reverted the change in ESLint for now, but it's possible we will apply the change again in a future major release, since it's useful for rule authors. To avoid this issue in the future, it would be nice if `espurify` ignored a `parent` property on nodes (or if it just ignored circular references in general) so that it wouldn't cause a stack overflow in this case.

**Comment by twada** [comment]
@not-an-aardvark Thank you for detailed explanation!
Okay I'm going to add circular references detection to `espurify`.

**Comment by novemberborn** [comment]
Thanks for digging into this @twada and @not-an-aardvark!

**Comment by gajus** [comment]
@not-an-aardvark Getting this issue again:

```
Maximum call stack size exceeded
RangeError: Maximum call stack size exceeded
    at cloneObject (/dev/surgeon/node_modules/espurify/lib/clone-ast.js:46:24)
    at cloneNodeOrObject (/dev/surgeon/node_modules/espurify/lib/clone-ast.js:21:14)
    at cloneOf (/dev/surgeon/node_modules/espurify/lib/clone-ast.js:64:16)
    at cloneObject (/dev/surgeon/node_modules/espurify/lib/clone-ast.js:52:20)
    at cloneNodeOrObject (/dev/surgeon/node_modules/espurify/lib/clone-ast.js:21:14)
    at cloneOf (/dev/surgeon/node_modules/espurify/lib/clone-ast.js:64:16)
    at cloneArray (/dev/surgeon/node_modules/espurify/lib/clone-ast.js:29:18)
    at cloneOf (/dev/surgeon/node_modules/espurify/lib/clone-ast.js:62:16)
    at cloneNode (/dev/surgeon/node_modules/espurify/lib/clone-ast.js:40:22)
    at cloneNodeOrObject (/dev/surgeon/node_modules/espurify/lib/clone-ast.js:19:14)

```

The issue can be replicated by running `eslint ./test` in https://github.com/gajus/surgeon project.

**Comment by not-an-aardvark** [comment]
This is probably because the change has been reapplied in ESLint v5.0 -- see the [migration guide](https://eslint.org/docs/5.0.0/user-guide/migrating-to-5.0.0#parent-before-rules).

**Comment by gajus** [comment]
> This is probably because the change has been reapplied in ESLint v5.0 -- see the migration guide.

Whats the relevance of the migration guide if the issue originates from espurify?

Actually, I was supposed to ping @twada in the first place. :-)

**Comment by not-an-aardvark** [comment]
Well, I think this plugin uses `espurify` as part of a rule, and `espurify` can cause a stack overflow if the AST has circular references. So in order to be compatible with ESLint v5, either `espurify` should add cycle detection, or this plugin should stop using `espurify` in ways that cause it to crash.

**Comment by twada** [comment]
Okay I'll fix it ASAP!

**Comment by twada** [comment]
FYI: [espurify v1.8.1](https://github.com/estools/espurify/releases/tag/v1.8.1) has been released.

**Comment by aginz** [comment]
Will it be possible to upgrade the package.json to the latest espurify?  Seems like ESLint v5 has caused  this error to occur again when "plugin:ava/recommended" is included.  After this plugin is removed, eslint works.

**Comment by novemberborn** [comment]
@aginz you'll have to do a deep update of the `eslint-plugin-ava` package. The SemVer range in our package does allow for `espurify@1.8.1`, but npm isn't very aggressive about updating transitive dependencies. Removing the dependency and then reinstalling it should also fix this.

Regardless, I'll land #208 and do a patch release which lists `^1.8.1` as its minimal dependency, so that should force the update.

**Comment by novemberborn** [comment]
> Regardless, I'll land #208 and do a patch release

I've published `5.5.1`.
